File: rust/src/protocol/v763/packets/play/clientbound/entity_equipment.rs
//! Packet `entity_equipment` (play/clientbound). Hand-written.

use crate::codec::uuid_codec::{self as uuid_c, Uuid};
use crate::codec::{
    bitset, chat_component, identifier, nbt, position, slot,
    string_codec as string, varint, varlong, BytesReader, BytesWriter, Reader, Writer,
};
use crate::errors::ProtocolError;
use crate::protocol::v763::states::ConnectionState;
use crate::protocol::v763::ClientboundPacket;

pub const PACKET_ID: i32 = 0x55;

#[derive(Debug, Clone, PartialEq, Default)]
pub struct EquipmentEntry {
    pub slot: i8,
    pub item: Option<slot::SlotData>,
}

#[derive(Debug, Clone, PartialEq, Default)]
pub struct EntityEquipment {
    pub entity_id: i32,
    pub equipments: Vec<EquipmentEntry>,
}

impl EntityEquipment {
    pub fn decode(reader: &mut BytesReader<'_>) -> Result<Self, ProtocolError> {
        let entity_id = varint::read(reader)?;
        let mut equipments = Vec::new();
        loop {
            let raw_slot = reader.read_exact(1)?[0];
            let more = raw_slot & 0x80 != 0;
            let mut s = (raw_slot & 0x7F) as i32;
            if s & 0x40 != 0 { s -= 0x80; }
            let item = slot::read(reader)?;
            equipments.push(EquipmentEntry { slot: s as i8, item });
            if !more { break; }
        }
        Ok(Self { entity_id, equipments })
    }
}

impl ClientboundPacket for EntityEquipment {
    fn state(&self) -> ConnectionState { ConnectionState::Play }
    fn packet_id(&self) -> i32 { PACKET_ID }
    fn encode(&self, writer: &mut BytesWriter) -> Result<(), ProtocolError> {
        varint::write(self.entity_id, writer)?;
        for (i, e) in self.equipments.iter().enumerate() {
            let more = i + 1 < self.equipments.len();
            let mut raw = (e.slot as u8) & 0x7F;
            if more { raw |= 0x80; }
            writer.write_all(&[raw])?;
            slot::write(e.item.as_ref(), writer)?;
        }
        Ok(())
    }
}

File: python/minecraft_bot/protocol/v763/packets/play/clientbound/entity_equipment.py
"""Packet `entity_equipment` (play/clientbound, id 0x55).

Updates an entity's worn/held items. The wire format uses a "top-bit
terminated array": each entry's slot byte's high bit (0x80) indicates
"more entries follow"; clearing the bit marks the last entry. Slot
codes 0-5 are: main hand, off hand, boots, leggings, chestplate, helmet.
"""

from __future__ import annotations

from dataclasses import dataclass

from minecraft_bot.codec import Reader, Writer, slot, varint

PACKET_ID = 0x55


@dataclass(frozen=True, slots=True)
class EquipmentEntry:
    slot: int                       # i8 (0..5); top bit reserved for stream
    item: slot.SlotData | None


@dataclass(frozen=True, slots=True)
class EntityEquipment:
    entity_id: int
    equipments: tuple[EquipmentEntry, ...]


def decode(reader: Reader) -> EntityEquipment:
    eid = varint.read(reader)
    entries: list[EquipmentEntry] = []
    while True:
        raw_slot = reader.read(1)[0]
        more = bool(raw_slot & 0x80)
        s = raw_slot & 0x7F
        if s & 0x40:
            s -= 0x80  # sign-extend lower 7 bits to i8
        item = slot.read(reader)
        entries.append(EquipmentEntry(slot=s, item=item))
        if not more:
            break
    return EntityEquipment(entity_id=eid, equipments=tuple(entries))


def encode(packet: EntityEquipment, writer: Writer) -> None:
    varint.write(packet.entity_id, writer)
    for i, entry in enumerate(packet.equipments):
        is_last = i == len(packet.equipments) - 1
        if not -64 <= entry.slot <= 63:
            from minecraft_bot.errors import ValueOutOfRange
            raise ValueOutOfRange("entity_equipment.slot", entry.slot)
        # Encode slot in low 7 bits (signed) plus continuation bit.
        b = (entry.slot & 0x7F) | (0x00 if is_last else 0x80)
        writer.write(bytes([b]))
        slot.write(entry.item, writer)
